infrastructure/smart-contracts/contracts/upgrades/esXai/esXai4.sol

```diff
@@ -0,0 +1,430 @@
+            if (list[i] == id) {
+                return i;
+            }
+        }
+        return list.length; // Return an invalid index if not found
+    }
+
+    /**
+    * @dev Removes a specific ID from a list of IDs.
+    * @param account The address of the account to remove the ID from.
+    * @param id The ID to remove from the list.
+    * @notice The function shifts elements to the left and pops the last element to shrink the array.
+    */
+    function removeIndexFromPendingRedemptionsList(address account, uint256 id) internal {
+        uint256 [] storage list = pendingRedemptionIds[account];
+        uint256 index = findIndexOfId(id, list);
+
+        require(index < list.length, "Value not found in array");
+
+        // Shift elements to the left
+        for (uint256 i = index; i < list.length - 1; i++) {
+            list[i] = list[i + 1];
+        }
+
+        // Remove the last element
+        list.pop();
+    }
+
+    /**
+    * @dev Calculates the total amount of pending redemptions for a given account.
+    * @param account The address of the account to calculate the total for.
+    * @return total The total amount of pending redemptions.
+    */
+    function getTotalesXaiPendingRedemptions(address account) public view returns (uint256 total) {
+        for (uint256 i = 0; i < pendingRedemptionIds[account].length; i++) {
+            total += _extRedemptionRequests[account][pendingRedemptionIds[account][i]].amount;
+        }
+    }
+
+    /**
+    * @notice Retrieves all pending redemption requests for a specific user.
+    * @dev This function returns an array of `RedemptionRequestExt` structs representing the pending 
+    *      redemption requests associated with the given account.
+    * @param account The address of the user whose pending redemption requests are to be retrieved.
+    * @return requests An array of `RedemptionRequestExt` structs representing the pending redemption requests.
+    */
+    function getPendingRedemptionsByUser(address account) external view returns (RedemptionRequestExt[] memory requests) {
+        uint256[] memory pendingIds = pendingRedemptionIds[account];
+        requests = new RedemptionRequestExt[](pendingIds.length);
+        for (uint256 i = 0; i < pendingIds.length; i++) {
+            requests[i] = _extRedemptionRequests[account][pendingIds[i]];
+        }
+    }
+
+    /**
+    * @dev Retrieves the most recently completed redemption requests for a given user, up to a maximum quantity.
+    * @param account The address of the user whose completed redemption requests are to be retrieved.
+    * @param maxQty The maximum number of completed redemption requests to return.
+    * @return requests An array of `RedemptionRequestExt` structs representing the most recent completed redemptions.
+    */
+    function getRecentlyCompletedRedemptionsByUser(address account, uint256 maxQty) external view returns (RedemptionRequestExt[] memory requests) {
```

Comment:
If we keep this function, should we add an offset param to facilitate pagination?